src/index.ts

```diff
@@ -1764,9 +1764,8 @@ export class Stream<T> implements InternalListener<T> {
    *
    * @return {Stream}
    */
-  flatten<R>(this: Stream<Stream<R>>): T {
-    const p = this._prod;
-    return new Stream<R>(new Flatten(this)) as T & Stream<R>;
+  flatten<R>(this: Stream<Stream<R>>): Stream<R> {
```

Comment:
I wonder what was the need for this change. I remember when I authored this line, that `T` was necessary, `Stream<R>` didn't suffice because `R` is a free variable (often the usage is `num$$.flatten()` not `num$$.flatten<number>()`), but it might be that it just works nowadays with the newest TS. Did you verify that this line *had* to be changed to satisfy compilation?

Reply:
I tried to get correct typehints here https://github.com/staltz/xstream/pull/268/files#diff-20c02845523c7301982ad63461750c4fR338.

When looking at the implementation of flatten, it always output a Stream, so T is plain wrong if `Stream<MemoryStream<A>>.flatten` (T being `MemoryStream<A>`) gives `MemoryStream<A>`. It should always give Stream<A>, whatever the combinations of Stream/MemoryStream. It didn't without this change.

Reply:
@staltz typescript infers those